### PR TITLE
feat: add hasStrategies and hasEnabledStrategies on feature environments

### PR DESCRIPTION
**Pull request by Tymek**
## About the changes
Trying to fetch all details about the feature needed for Project Overview, to remove `useFeature` from `FeatureToggleSwitch`.

- decouple feature toggle switch from `useFeature` - prevent fetching for every row
- move modals from table cell level into overview, improving performance
- rewrite logic into `onToggle` - middleware pattern on action

### CODE REVIEW

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/useFeatureToggleSwitch.tsx [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+                handleToggleEnvironmentOff,
+                handleToggleEnvironmentOn,
+            ]);
+        },
+        [setProdGuardModalState],
+    );
+
+    const modals = (
+        <>
+            <FeatureStrategyProdGuard {...prodGuardModalState} />
```
Comment (FredrikOseberg) [inline frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/useFeatureToggleSwitch.tsx:210]: The checked state of the `Don't show again` checkbox seems to be broken

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+            ) {
+                onError?.();
+            } else {
+                setToastApiError(formatUnknownError(error));
+            }
+            rollbackIsChecked();
+        }
+    };
+
+    const handleToggleEnvironmentOff = async () => {
```
Comment (FredrikOseberg) [inline frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx:108]: Should follow the ADR of `onAction` for handlers

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        environmentName,
+    );
+    const [showProdGuard, setShowProdGuard] = useState(false);
+
+    const disabledStrategiesCount =
+        feature?.environments
+            .find((env) => env.name === environmentName)
+            ?.strategies.filter((strategy) => strategy.disabled).length ?? 0;
+
+    const handleToggleEnvironmentOn = async (
```
Comment (FredrikOseberg) [inline frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx:78]: Should follow the ADR of `onAction` for handlers

File: src/lib/openapi/spec/feature-environment-schema.ts [hunk 4]
```diff
@@ -71,6 +71,15 @@ export const featureEnvironmentSchema = {
...
         },
+        hasStrategies: {
+            type: 'boolean',
+            description: 'Whether the feature has any strategies defined.',
+        },
+        hasEnabledStrategies: {
+            type: 'boolean',
+            description:
+                'Whether the feature has any enabled strategies defined.',
+        },
```
Comment (FredrikOseberg) [inline src/lib/openapi/spec/feature-environment-schema.ts:82]: Where are the tests verifying that these are part of the payload?

**Review by FredrikOseberg — COMMENTED** [review]
This change is massive. Are there tests covering this page? If not we should definitely add them.

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        if (featureHasOnlyDisabledStrategies()) {
+            setShowEnabledDialog(true);
+        } else {
+            await handleToggleEnvironmentOn();
+        }
+    };
+
+    const onClick = async () => {
+        if (enableProdGuard && !isChangeRequestConfigured(environmentName)) {
+            setShowProdGuard(true);
```
[the remark this answers is not in the document]
Comment (Tymek) [reply frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx:157]: This is in `LegacyFeatureToggle` - copy of current code, not the modified part that is behind the flag

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx (hunk 3 above)
Comment (Tymek) [reply frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx:78]: `LegacyFeatureToggleSwitch.tsx`, not new code

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/useFeatureToggleSwitch.tsx (hunk 1 above)
Comment (Tymek) [reply frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/useFeatureToggleSwitch.tsx:210]: fixed

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx (hunk 2 above)
Comment (Tymek) [reply frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/LegacyFeatureToggleSwitch.tsx:108]: `LegacyFeatureToggleSwitch.tsx`, not new code

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/createFeatureToggleCell.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+}));
+
+export const createFeatureToggleCell =
+    (
+        projectId: string,
+        environmentName: string,
+        isChangeRequestEnabled: boolean,
+        refetch: () => void,
+        onFeatureToggleSwitch: ReturnType<UseFeatureToggleSwitchType>['onToggle'],
+    ) =>
```
Comment (Tymek) [inline frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/createFeatureToggleCell.tsx:32]: simplify component factory arguments type

File: frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/FeatureToggleSwitch.types.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+export type OnFeatureToggleSwitchArgs = {
+    featureId: string;
+    projectId: string;
+    environmentName: string;
+    environmentType?: string;
+    hasStrategies?: boolean;
+    hasEnabledStrategies?: boolean;
+    isChangeRequestEnabled?: boolean;
+    onRollback?: () => void;
```
Comment (Tymek) [inline frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/FeatureToggleSwitch.types.ts:11]: `onCancel`

**Review by kwasniew — APPROVED** [review]
Mob reviewed

**Pull request merged**